### PR TITLE
Fix: do not fail and exit properly on errors

### LINKED ISSUE #3: UnhandledPromiseRejectionWarning: TypeError: Cannot read property '0' of null
When I try to transform the following tags:

```html
<meta property="og:title" content="V!RTL" />
<meta property="og:image" content="assets/logo.svg" />
<meta property="og:locale" content="en_US" />
<meta property="og:site_name" content="v1rtl.site" />
```

I get this error:

```
UnhandledPromiseRejectionWarning: TypeError: Cannot read property '0' of null
    at getMetaTag (/home/v1rtl/Desktop/Coding/v1rtl.site/node_modules/parcel-plugin-ogimage/index.js:11:25)
    at Bundler.<anonymous> (/home/v1rtl/Desktop/Coding/v1rtl.site/node_modules/parcel-plugin-ogimage/index.js:35:20)
    at Bundler.emit (events.js:203:13)
    at Bundler.bundle (/home/v1rtl/Desktop/Coding/v1rtl.site/node_modules/parcel-bundler/src/Bundler.js:361:12)
    at processTicksAndRejections (internal/process/task_queues.js:85:5)
(node:15421) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:15421) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

Assets content:

```sh
➜ tree assets -L 1
assets
├── animation
├── art_bg.webp
├── background.webp
├── blog.webp
├── desktop.webp
├── icons
├── logo.svg
├── marker.webp
├── robots.webp
└── web_bg.webp
```

### PR DESCRIPTION
**Pull request by JPeer264**
Closes #3

Screenshot 2019-08-15 at 10 42 55

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -5,46 +5,69 @@ const ora = require('ora');
 const chalk = require('chalk');
 const prettyMs = require('pretty-ms');
 
-const getMetaTag = (html, property) => {
+const getMetaTag = (html, property, spinner) => {
```
Comment (lukechilds) [inline index.js]: Same as with `getMetaTagContent`.

File: index.js [hunk 2]
```diff
@@ -5,46 +5,69 @@ const ora = require('ora');
...
+		spinner.fail(`Missing ${property}`);
+
+		return;
+	}
+
+	return executedRegex[0];
 };
 
-const getMetaTagContent = metaTagHtml => {
+const getMetaTagContent = (metaTagHtml, property, spinner) => {
```
Comment (lukechilds) [inline index.js]: We shouldn't be passing property/spinner through here and manipulating the spinner inside the funtion. Instead the function should throw/return false and then handle that result elsewhere in the code.

File: index.js [hunk 3]
```diff
@@ -5,46 +5,69 @@ const ora = require('ora');
...
-	return regex.exec(metaTagHtml)[1];
+	return executedRegex[1];
 };
 
 module.exports = bundler => {
 	bundler.on('buildEnd', async () => {
 		if (process.env.NODE_ENV !== 'production') {
 			return;
 		}
-		console.log('');
```
Comment (lukechilds) [inline index.js:25]: Why is this removed?

**Review by lukechilds — APPROVED** [review]
Thanks for working on this.

The code can be made much more modular and simpler by throwing an error in the failing functions and then wrapping the existing code in a try/catch that properly handles that error.

**Review by lukechilds — CHANGES_REQUESTED** [review]
Thanks for working on this.

The code can be made much more modular and simpler by throwing an error in the failing functions and then wrapping the existing code in a try/catch that properly handles that error.

File: index.js (hunk 3 above)
Comment (JPeer264) [reply index.js:25]: Whoops, good catch

**Comment by JPeer264** [comment]
I wrapped now the code in a `try/catch` - This handles just one error at a time though. You think that is good like that?

**Comment by JPeer264** [comment]
@lukechilds ping

File: index.js [hunk 4]
```diff
@@ -7,14 +7,24 @@ const prettyMs = require('pretty-ms');
...
-	return regex.exec(html)[0];
+	if (!executedRegex) {
+		throw new Error(`Missing ${property}`);
+	}
+
+	return executedRegex[0];
 };
 
-const getMetaTagContent = metaTagHtml => {
+const getMetaTagContent = (metaTagHtml, property) => {
```
Comment (lukechilds) [inline index.js]: Still no need to be passing property in here.

We can grab it from the `metaTagHtml`. Or even just dump the entire contents of `metaTagHtml` to the console. It's probably not a lot of text.

File: index.js [hunk 5]
```diff
@@ -7,14 +7,24 @@ const prettyMs = require('pretty-ms');
 
 const getMetaTag = (html, property) => {
 	const regex = new RegExp(`<meta[^>]*property=["|']${property}["|'][^>]*>`, 'i');
+	const executedRegex = regex.exec(html);
```
Comment (lukechilds) [inline index.js]: I think results is a better name for this variable.

Comment (lukechilds) [inline index.js]: I think results is a better name for this variable.

File: index.js (hunk 4 above)
Comment (lukechilds) [inline index.js]: Still no need to be passing property in here.

We can grab it from the `metaTagHtml`. Or even just dump the entire contents of `metaTagHtml` to the console. It's probably not a lot of text.

File: index.js [hunk 6]
```diff
@@ -23,28 +33,41 @@ module.exports = bundler => {
...
-		fs.writeFileSync(htmlPath, patchedHtml);
+			fs.writeFileSync(htmlPath, patchedHtml);
+		} catch (error) {
+			spinner.fail(error.message);
+
+			hasErrors = true;
+		}
 
 		const end = Date.now();
+		const symbol = hasErrors ? chalk.red('✖') : '✨ ';
```
Comment (lukechilds) [inline index.js:63]: Do we need a space in `'✖'` here to keep the style consistent with the `'✨ '` symbol?

File: index.js [hunk 7]
```diff
@@ -23,28 +33,41 @@ module.exports = bundler => {
...
+		const text = hasErrors ?
+			chalk.red('Failed to fix og:image link. Please fix the above error.') :
```
Comment (lukechilds) [inline index.js]: Maybe remove the `Please fix the above error.` text here.

This could be triggered by `fs.writeFileSync` and it might not make sense to say fix whatever the output of that is.

Just `Failed to fix og:image link.` should be fine. If there is a clear error above it the user will know that's what needs attention.

**Review by lukechilds — CHANGES_REQUESTED** [review]
@JPeer264 sorry for the delay.

>I wrapped now the code in a try/catch - This handles just one error at a time though. You think that is good like that?

Yes that's perfect. If an error is encountered we should fail early. It's not normally a good idea to keep on working on something if something's gone wrong.

File: index.js (hunk 6 above)
Comment (JPeer264) [reply index.js:63]: In this case it looks better without a space, as `spinner.fail` does not have a space either:

Screenshot 2020-03-19 at 17 20 38

**Comment by JPeer264** [comment]
@lukechilds I updated the PR.

**Comment by lukechilds** [comment]
Thanks @JPeer264.

Closed in https://github.com/lukechilds/parcel-plugin-ogimage/commit/6d70ba0e821295862586aab4204e9773d27454c5.
